File: app/src/main/java/com/anenha/weather/app/utils/Formatter.java
package com.anenha.weather.app.utils;

import android.content.Context;
import com.anenha.weather.R;
import com.anenha.weather.app.model.Condition;
import com.anenha.weather.app.model.Location;
import com.anenha.weather.app.provider.TranslateCallback;
import com.anenha.weather.app.provider.TranslateService;

/**
 * Created by ajnen on 12/10/2017.
 */

public class Formatter {

    public static String location(final Location location) {
        return location.getCity() + ", " + location.getCountry();
    }

    public static String temperature(final Integer temperture, final String tempUnit) {
        switch (tempUnit) {
            case "1": return temperture + "\u00B0" + "C";
            case "2": return celsiusToFahrenheit(temperture);
            default: return temperture + "\u00B0" + "C";
        }
    }

    public static String time(final String hour, final String timeFormat) {
        switch (timeFormat) {
            case "1": return getTime(hour);
            case "2": return hour;
            default: return hour;
        }
    }

    public static String weekDay(final Context context, final String weekDay) {
        return getWeekday(context, weekDay, true);
    }

    public static String date(final Context context, final String day, final String date, final boolean monthInitials, final boolean hasYear) {
        return getDate(context, day, date, monthInitials, hasYear);
    }

    public static String conditionDescription(final Context context, final Condition condition){
        final String[] conditionsArray = context.getResources().getStringArray(R.array.condition_descriptions);
        final Integer code = Integer.parseInt(condition.getCode());
        if (code >= 0 && code <= 47) {
            return conditionsArray[code];
        } else {
            return context.getString(R.string.condition_unavailable);
        }
    }

    private static String celsiusToFahrenheit(final Integer celsusTemp){
        return (celsusTemp *9/5 + 32)+ "\u00B0" + "F";
    }

    private static String getTime(String time) {
        String[] hour = time.split(" ");

        if(hour[1].equalsIgnoreCase("AM")){
            return "0" + hour[0];
        } else {
            String[] time_PM = hour[0].split(":");
            int afternoon = Integer.parseInt(time_PM[0]) + 12;
            return afternoon + ":" + time_PM[1];
        }
    }

    private static String getDate(final Context context, final String day, final String date, final boolean monthInitials,  final boolean hasYear){
        String formatedDate = "";

        if(day != null && !day.isEmpty()){
            formatedDate = getWeekday(context, day, false) + ", ";
        }

        final String[] finalDate = date.split(" ");
        if(Integer.parseInt(finalDate[0]) < 10){ finalDate[0] = "0"+finalDate[0]; }
        finalDate[1] = getMonth(context, finalDate[1], monthInitials);
        formatedDate = formatedDate + finalDate[0] + " " + finalDate[1];

        if(hasYear){ formatedDate = formatedDate + " " + finalDate[2]; }
        return formatedDate;
    }

    private static String getWeekday(final Context context, final String weekday, final boolean initials){
        final String[] weekdays = context.getResources().getStringArray(R.array.weekdays);
        String day = "";

        switch (weekday){
            case "Sun": day = weekdays[0]; break;
            case "Mon": day = weekdays[1]; break;
            case "Tue": day = weekdays[2]; break;
            case "Wed": day = weekdays[3]; break;
            case "Thu": day = weekdays[4]; break;
            case "Fri": day = weekdays[5]; break;
            case "Sat": day = weekdays[6]; break;
        }

        if(initials){ return day.substring(0, 3); }
        return day;
    }

    private static String getMonth(final Context context, final String key, final boolean initials){
        final String[] months = context.getResources().getStringArray(R.array.months);
        String month = "";

        switch (key){
            case "Jan": month = months[0]; break;
            case "Feb": month = months[1]; break;
            case "Mar": month = months[2]; break;
            case "Apr": month = months[3]; break;
            case "May": month = months[4]; break;
            case "Jun": month = months[5]; break;
            case "Jul": month = months[6]; break;
            case "Aug": month = months[7]; break;
            case "Sep": month = months[8]; break;
            case "Oct": month = months[9]; break;
            case "Nov": month = months[10]; break;
            case "Dec": month = months[11]; break;
        }

        if(initials){ return month.substring(0, 3); }
        return month;
    }
}

File: app/src/main/java/com/anenha/weather/app/utils/Formatter.kt
package com.anenha.weather.app.utils

import android.content.Context
import com.anenha.weather.R
import com.anenha.weather.app.model.ConditionModel
import com.anenha.weather.app.model.LocationModel

import java.text.NumberFormat

/**
 * Created by ajnen on 12/10/2017.
 */

object Formatter {

    fun location(location: LocationModel, fullLocation: Boolean): String {
        val locale = StringBuilder()
        locale.append(location.city).append(", ")
        if (fullLocation) {
            locale.append(location.region!!.trim { it <= ' ' }).append(", ")
        }
        locale.append(location.country)
        return locale.toString()
    }

    fun temperature(temperture: Int?, tempUnit: String): String {
        when (tempUnit) {
            "1" -> return temperture.toString() + "\u00B0" + "C"
            "2" -> return celsiusToFahrenheit(temperture)
            else -> return temperture.toString() + "\u00B0" + "C"
        }
    }

    fun time(hour: String, timeFormat: String): String {
        when (timeFormat) {
            "1" -> return getTime(hour)
            "2" -> return hour
            else -> return hour
        }
    }

    fun weekDay(context: Context, weekDay: String): String {
        return getWeekday(context, weekDay, true)
    }

    fun date(context: Context, day: String, date: String, monthInitials: Boolean, hasYear: Boolean): String {
        return getDate(context, day, date, monthInitials, hasYear)
    }

    fun conditionDescription(context: Context, condition: ConditionModel): String {
        val conditionsArray = context.resources.getStringArray(R.array.condition_descriptions)
        val code = Integer.parseInt(condition.code)
        return if (code >= 0 && code <= 47) {
            conditionsArray[code]
        } else {
            context.getString(R.string.condition_unavailable)
        }
    }

    private fun celsiusToFahrenheit(celsusTemp: Int?): String {
        return (celsusTemp!! * 9 / 5 + 32).toString() + "\u00B0" + "F"
    }

    private fun getTime(time: String): String {
        val hour = time.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        if (hour[1].equals("AM", ignoreCase = true)) {
            return "0" + hour[0]
        } else {
            val time_PM = hour[0].split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val afternoon = Integer.parseInt(time_PM[0]) + 12
            return afternoon.toString() + ":" + time_PM[1]
        }
    }

    private fun getDate(context: Context, day: String?, date: String, monthInitials: Boolean, hasYear: Boolean): String {
        var formatedDate = ""

        if (day != null && !day.isEmpty()) {
            formatedDate = getWeekday(context, day, false) + ", "
        }

        val finalDate = date.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        val intFormater = NumberFormat.getIntegerInstance()
        intFormater.minimumIntegerDigits = 2
        intFormater.maximumIntegerDigits = 2
        intFormater.isGroupingUsed = false
        val monthDay = intFormater.format(Integer.parseInt(finalDate[0]).toLong())

        finalDate[1] = getMonth(context, finalDate[1], monthInitials)
        formatedDate = formatedDate + monthDay + " " + finalDate[1]

        if (hasYear) {
            formatedDate = formatedDate + " " + finalDate[2]
        }
        return formatedDate
    }

    private fun getWeekday(context: Context, weekday: String, initials: Boolean): String {
        val weekdays = context.resources.getStringArray(R.array.weekdays)
        var day = ""

        when (weekday) {
            "Sun" -> day = weekdays[0]
            "Mon" -> day = weekdays[1]
            "Tue" -> day = weekdays[2]
            "Wed" -> day = weekdays[3]
            "Thu" -> day = weekdays[4]
            "Fri" -> day = weekdays[5]
            "Sat" -> day = weekdays[6]
        }

        return if (initials) {
            day.substring(0, 3)
        } else day
    }

    private fun getMonth(context: Context, key: String, initials: Boolean): String {
        val months = context.resources.getStringArray(R.array.months)
        var month = ""

        when (key) {
            "Jan" -> month = months[0]
            "Feb" -> month = months[1]
            "Mar" -> month = months[2]
            "Apr" -> month = months[3]
            "May" -> month = months[4]
            "Jun" -> month = months[5]
            "Jul" -> month = months[6]
            "Aug" -> month = months[7]
            "Sep" -> month = months[8]
            "Oct" -> month = months[9]
            "Nov" -> month = months[10]
            "Dec" -> month = months[11]
        }

        return if (initials) {
            month.substring(0, 3)
        } else month
    }
}
